File: 25-07/shop-frontend/src/app/services/user-service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface User {
  userId: number;
  username: string;
  password?: string; // Only used during create/edit
}

export interface UserDTO {
  username: string;
  password: string;
}

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private apiUrl = 'http://localhost:5043/api/User';

  constructor(private http: HttpClient) {}

  getAll(): Observable<User[]> {
    return this.http.get<User[]>(this.apiUrl);
  }

  getById(id: number): Observable<User> {
    return this.http.get<User>(`${this.apiUrl}/Details/${id}`);
  }

  create(user: UserDTO): Observable<User> {
    return this.http.post<User>(`${this.apiUrl}/Create`, user);
  }

  update(id: number, user: UserDTO): Observable<User> {
    return this.http.post<User>(`${this.apiUrl}/Edit/${id}`, user);
  }

  delete(id: number): Observable<User> {
    return this.http.post<User>(`${this.apiUrl}/Delete/${id}`, {});
  }
  
}

File: 25-07/Migration/ClientVHShop/Services/UserService.cs
using BCrypt.Net;
using ChienVHShopOnline.DTOs;
using ChienVHShopOnline.Models;
using ChienVHShopOnline.Repositories;
using Microsoft.AspNetCore.Mvc;

namespace ChienVHShopOnline.Services;

public class UserService
{
    private UserRepo _userRepo;
    public UserService(UserRepo UserRepo)
    {
        _userRepo = UserRepo;
    }

    public async Task<List<User>> GetAll()
    {
        return (await _userRepo.GetAll()).ToList();
    }

    public async Task<User> Get(int id)
    {
        return (await _userRepo.Get(id));
    }

    public async Task<User> Create([FromBody] UserDTO userDTO)
    {
        var User = new User
        {
            Username = userDTO.Username,
            Password = BCrypt.Net.BCrypt.EnhancedHashPassword(userDTO.Password)
        };
        User = await _userRepo.Add(User);
        return User;
    }
    public async Task<User> Edit(int id, [FromBody] UserDTO userDTO)
    {
        User user = await _userRepo.Get(id);
        user.Username = userDTO.Username;
        user.Password = BCrypt.Net.BCrypt.EnhancedHashPassword(userDTO.Password);


        user = await _userRepo.Update(id, user);
        return user;
    }
    public async Task<User> Delete(int id)
    {
        User user = await _userRepo.Delete(id);
        return user;
    }
    
}